src/Microsoft.Management.Infrastructure.CimCmdlets/CimAsyncOperation.cs

```diff
@@ -443,11 +442,11 @@ protected bool Disposed
         {
             get
             {
-                return Interlocked.Read(ref this._disposed) == 1;
+                return Interlocked.CompareExchange(ref this._disposed, 0, 0) == 1;
```

Comment:
I am not sure why it is so important to do atomic read on each get.  But this is old code and I'd rather not make logic changes.  It seems like _disposed should be tagged 'volatile' if it is that sensitive.

Anyway, what I don't like about this is that most of the time _disposed is '0' and so presumably each get will result in a write operation.  It would be better to test for the final value of '1', which only happens once, and avoid multiple writes.

```
return Interlocked.CompareExchange(ref this._disposed, 1, 1) == 1;
```

Reply:
@PaulHigin that code will overwrite the value with `1` on the first read.

Reply:
Replacement only occurs if the value and comparison value are equal.  The _disposed value is '0' and the comparison value is '1'.  

But this illustrates a serious downside to this change.  It is difficult to read and reason about, and so difficult to maintain.  There is not a large upside with this change, so I feel it should not be taken.

Reply:
Interlocked APIs are never fun to reason about. An additional comment should be added, I feel, and arguments named for clarity, but the change itself is sound IMO.

Reply:
@PaulHigin I don't think the atomic read using `CompareExchange` is necessary, it is just being used to ensure a full fence. `volatile` does not protect against a read after write so it is no use here.

Reply:
It is not just using Interlocked APIs, but using them in an unusual way.  This makes the code hard to read and maintain.  I agree that the atomic read is probably not necessary, but I also see no great harm in it.  It is old code and there is no real upside in changing it.  Therefore I won't approve the change and instead suggest the code be left as is.

/cc @daxian-dbw

Reply:
I had a think about this and we can eliminate the dummy `CompareExchange` by using an explicit memory barrier (see 7fc1f28). This no longer relies upon side effects from `CompareExchange`.